sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/python/GpuArrowEvalPythonExec.scala

```diff
@@ -170,63 +170,154 @@ class RebatchingRoundoffIterator(
   }
 }
 
+/** A listener to work with the BatchQueue. */
+trait BatchQueuePeekListener {
```

Comment:
Can we have a follow on issue to come back and try to remove the custom traits that are extending Iterators to not act like Iterators any more. I don't like Iterators that don't behave like a real iterator and need special calls to make them behave properly.  Either we need to make it so it is not an Iterator, or it is an iterator and does not need special calls to keep it working properly.

Reply:
I updated it in this PR since it targets to 24.02 now.

Reply:
I introduced `BatchProducer` again to support the peek operation for `BatchQueue`, replacing the listener things. The `BatchProducer` accepts an iterator as the input and supports accessing the batches from the input by a new iterator returned from `asIterator` method. 
Now `BatchProducer` is not an iterator, but it produces an pure iterator.